File: algorithm1/chaptet_7/fence.py
def fence(height, start, end):
    if(start == end):
        return height[start]
    mid = int((start+end)/2)
    ret = max(fence(height,start,mid), fence(height, mid+1,end))
    lo = mid
    hi = mid+1
    h = min(height[lo], height[hi])
    ret = max(ret, h*2)
    while(start < lo or hi < end):
        if(hi<end) and ((lo == start) or (height[lo-1]<height[hi+1])):
            hi = hi+1
            h = min(h, height[hi])
        else:
            lo = lo-1
            h = min(h, height[lo])
        ret = max(ret, h*(hi-lo+1))
    return ret

if __name__=="__main__":
    height = [7,1,5,9,6,7,3]
    res = fence(height, 0, 6)
    print(res)

File: chapter7/FENCE.cpp
#include <iostream>
#include <vector>

using namespace std;

int findMax(vector<int> left, vector<int> right) {
    int height;
    int width = 0;
    if(left[left.size()-1] < right[0]) height = left[left.size()-1];
    else height = right[0];

    int start = left.size()-1;
    while(1) {
        if(left[start] >= height) {
            width++;
            start = start - 1;
        } else break;
    }

    int start2 = 0;
    while(1) {
        if(right[start2] >= height) {
            width++;
            start2 = start2 +1;
        } else break;
    }
    return height * width;

}

int divide(vector<int> fence) {
    
    int len = fence.size();
    if(len == 1) return fence[0];
    vector<int> leftFence;
    vector<int> rightFence;
    int max = 0;
    for(int i = 0; i < len/2 ; i++) {
        leftFence.push_back(fence[i]);
    } 
    for(int i = len / 2; i < len; i++) {
        rightFence.push_back(fence[i]);
    }

    int maxLeft = divide(leftFence);
    if(maxLeft > max) max = maxLeft;
    int maxRight = divide(rightFence);
    if(maxRight > max) max = maxRight;
    int maxMiddle = findMax(leftFence, rightFence);
    if(maxMiddle > max) max = maxMiddle;
    return max;
}

int main() {
    vector<int> fence = {1,4,4,4,4,1,1};
    cout<< divide(fence) << endl;
}